definitions.h

```diff
@@ -51,6 +54,143 @@ struct Vertex
     double w;
 };
 
+/****************************************************
+* My Matrix class
+*****************************************************/
+class Matrix {
+
+public:
+
+   int height = 0;
+   int width = 0;
+// Create a Generic Matrix, this will be useful later
+   double matrix[4][4] = { {1,0,0,0},
+                           {0,1,0,0},
+                           {0,0,1,0},
+                           {0,0,0,1} };
+
+// Some constructors
+   Matrix(){
+   }
+
+   Matrix(double fbf[4][4]){
+      height = 4;
+      width = 4;
+      
+      for (int i = 0; i < 4; i++)
+      {
+         for (int j = 0; j < 4; j++)
+         {
+            matrix[i][j] = fbf[i][j];
+         }        
+      }
+         
+   }
+
+// Initialise with size only.
+   Matrix(int h, int w){
+      this->height = h;
+      this->width = w;
+   }
+
+// Multiplication operator overload for matrix operations
+   Matrix operator*(const Matrix &rhs)
+   {
+      //check if you can actually multiply them
+      if(width == rhs.height)
+      { 
+         Matrix res(rhs.height,rhs.width);
+
+          int i, j, k; 
+            for (i = 0; i < height; i++) 
+            { 
+               for (j = 0; j < rhs.width; j++) 
+               { 
+                     res.matrix[i][j] = 0; 
+                     for (k = 0; k < rhs.height; k++) 
+                        res.matrix[i][j] += this->matrix[i][k] *  
+                                    rhs.matrix[k][j]; 
+               } 
+            }
+         return res;
+      }
+   }
+
+// For shader operations
+   Matrix operator*(const Vertex &rhs)
+ 	{
+      Matrix res(4,1);
+
+      if(width == 4)
+      { 
+         double vertex[4][1] = { {rhs.x}, 
+                                 {rhs.y}, 
+                                 {rhs.z}, 
+                                 {rhs.w} };
+
+         for(int i = 0; i < height; i++){	
+            for(int j = 0; j < 1; j++){ 
```

Comment:
The second loop here is actually not needed, you are starting at 0 and looping until it's less than 1, it could be replaced by <code>res.matrix[i][0] += this->matrix[i][k] * vertex[k][0];</code>. In the end, it does the same thing it just might be more obvious that this is a specific kind of multiplication going on.